File: modules/common/src/models/colorSetting.ts
import { z } from "zod";

export const colorTargetTypes = ["Item", "ItemType"] as const;
export const colorScreens = ["master", "serve"] as const;

export const colorSettingSchema = z.object({
  id: z.string().uuid().optional(),
  target_type: z.enum(colorTargetTypes),
  target_id: z.string().uuid(),
  screen: z.enum(colorScreens),
  color: z.string().regex(/^#[0-9a-fA-F]{6}$/, "色は #RRGGBB で指定します"),
});

export type ColorSetting = z.infer<typeof colorSettingSchema>;
export type ColorTargetType = ColorSetting["target_type"];
export type ColorScreen = ColorSetting["screen"];

type ColorTarget = {
  id?: string;
  item_type: { id?: string };
};

export const findColorSetting = <T extends ColorSetting>(
  settings: T[],
  targetType: ColorTargetType,
  targetId: string | undefined,
  screen: ColorScreen,
): T | undefined =>
  settings.find(
    (setting) =>
      setting.target_type === targetType &&
      setting.target_id === targetId &&
      setting.screen === screen,
  );

/**
 * アイテムの背景色を設定から引く
 * item の設定 > item_type の設定の順で探し、どちらも無ければ undefined を返す。
 * undefined のときは呼び出し側の既定の色を使う。
 */
export const resolveItemColor = (
  settings: ColorSetting[],
  item: ColorTarget,
  screen: ColorScreen,
): string | undefined => {
  if (item.id) {
    const itemSetting = findColorSetting(settings, "Item", item.id, screen);
    if (itemSetting) return itemSetting.color;
  }
  if (item.item_type.id) {
    const typeSetting = findColorSetting(
      settings,
      "ItemType",
      item.item_type.id,
      screen,
    );
    if (typeSetting) return typeSetting.color;
  }
  return undefined;
};

File: api/internal/models/color_setting.go
// api/internal/models/color_setting.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// ColorSetting はマスター・提供画面でのアイテムの背景色設定。
//
// 画面ごとに1行持つ。同じ対象・同じ画面に2つの色が付かないよう、
// (target_type, target_id, screen) に一意制約を張っている。
// 論理削除にすると一意制約とぶつかるので、設定解除は物理削除にする。
type ColorSetting struct {
	ID         uuid.UUID `gorm:"type:uuid;primary_key;default:uuid_generate_v4()"`
	TargetType string    `gorm:"not null;uniqueIndex:idx_color_settings_target_screen"` // "Item" または "ItemType"
	TargetID   uuid.UUID `gorm:"type:uuid;not null;uniqueIndex:idx_color_settings_target_screen"`
	Screen     string    `gorm:"not null;uniqueIndex:idx_color_settings_target_screen"` // "master" または "serve"
	Color      string    `gorm:"not null"`                                              // #RRGGBB
	CreatedAt  time.Time `gorm:"not null"`
	UpdatedAt  time.Time `gorm:"not null"`
}

func (colorSetting *ColorSetting) BeforeCreate(tx *gorm.DB) error {
	if colorSetting.ID == uuid.Nil {
		colorSetting.ID = uuid.New()
	}
	return nil
}
